### DISCUSSION TITLE
How to clone a VRM model?

### DISCUSSION THREAD

**Discussion by mattrossman** [pixiv/three-vrm#1172] [Help]
In Three.js you can reuse glTF models by cloning them prior to adding them to the scene. Adding multiple copies of the same `gltf.scene` instance can otherwise lead to unexpected behavior.

For simple models this can be accomplished via `gltf.scene.clone()`. For models with skinned meshes, more care must be taken to clone internal objects. In these cases, you may clone the model using `SkeletonUtils` from `three/examples`:
```ts
const copy = SkeletonUtils.clone(gltf.scene)
scene.add(copy)
```

See discussion: https://discourse.threejs.org/t/how-to-clone-a-model-thats-loaded-with-gltfloader/23723/8

---

As VRM is an extension of glTF, I expect to be able to clone VRM models in a similar way as regular glTF models. However, while I can clone the `gltf.scene` with `SkeletonUtils.clone()` as shown above, this doesn't clone the `VRM` object associated with it, so I lose out on VRM functionality.

**Example code**: https://jsfiddle.net/mattrossman/satvh5kr/3/

As you can see in the example, I can clone and reposition the model, but I don't have a `VRM` object for the copy.

Is there a suggested way to fully clone VRM models? I see a [VRMHumanoid.clone()](https://pixiv.github.io/three-vrm/packages/three-vrm/docs/classes/VRMHumanoid.html#clone) method in the API reference, but I'm not sure if/how it would be used here.

**Comment by 0b5vr** [comment]
Cloning the entire VRM! I have never thought of it.
The current design of three-vrm definitely doesn't imagine that.

What is your purpose here?

**Reply by mattrossman** [reply]
My purpose is using this library with [React Three Fiber](https://github.com/pmndrs/react-three-fiber). The standard behavior for loaders in React Three Fiber is to cache assets in memory to improve performance: https://docs.pmnd.rs/react-three-fiber/api/hooks#useloader

This means that if you add the same model multiple times in a scene and/or plan to modify instances of the model at runtime, you need to clone it prior to use ([see issue thread](https://github.com/pmndrs/react-three-fiber/issues/245)). Note that this pattern is not unique to React though, the caching / cloning use case appears in vanilla Three as well ([discussion 1](https://discourse.threejs.org/t/solved-issue-with-gltfloader-and-reusing-geometry/6697), [discussion 2](https://discourse.threejs.org/t/is-there-anyway-i-can-cache-models-in-the-users-browser-cache/9132/12))


The React Three Fiber community has built a tool [gltfjsx](https://github.com/pmndrs/gltfjsx) to assist with reusing glTF assets, but that only works if you know your model's structure ahead of time. For arbitrary assets loaded at runtime you still need to do a deep clone for reuse.

Currently this library is structured as a GLTFLoader plugin which I register in [useGLTF](https://github.com/pmndrs/drei#usegltf) like so:
```ts
const gltf = useGLTF('/model.vrm', undefined, undefined, (loader) => {
  loader.register(parser => new VRMLoaderPlugin(parser, { autoUpdateHumanBones: true }))
})
```
This means that any time I load `'/model.vrm'` in my scene, I get the same `GLTF` instance with the same `VRM` instance, which causes hard-to-debug problems once I start manipulating VRM bones and retargeting animations.

Typically I would clone a glTF character asset with [SkeletonUtils](https://threejs.org/docs/?q=skele#examples/en/utils/SkeletonUtils) like so:

```tsx
/** @type {THREE.Object3D} */
const sceneCopy = useMemo(() => SkeletonUtils.clone(gltf.scene), [gltf.scene])
```

But as mentioned this doesn't give me access to a corresponding `VRM` object, just the plain skinned meshes and bones.

What I want is is something like:

```tsx
const vrm = gltf.userData.vrm

/** @type {VRM} */
const vrmCopy = useMemo(() => VRMUtils.clone(vrm), [vrm])
```
Then I could safely add the cloned `vrmCopy.scene` to my scene and manipulate it with `vrmCopy.humanoid`

**Reply by 0b5vr** [reply]
I see! That sounds like a legit motivation.
I'm still determining when we can work on this; I will discuss it with the team.

(I'm sorry for the late response :bow:)

**Comment by Samsy** [comment]
+1

**Comment by a-saq** [comment]
+1

**Reply by gwendall** [reply]
+1

**Comment by Ending2015a** [comment]
Any updates?

**Comment by saori-eth** [comment]
+1

**Comment by nionis** [comment]
One (not performant) workaround it to use different urls to load the same model: `/models/avatar.vrm`, `/models/avatar.vrm?x=1`, `/models/avatar.vrm?x=2` etc

**Comment by Benjythebee** [comment]
+1

**Comment by saori-eth** [comment]
world peace can be achieved if yall enable vrm cloning 🙏 

![image](https://github.com/user-attachments/assets/7de0e349-3a65-4f3f-bb69-35c0f6bd9dba)

**Comment by bagusindrayana** [comment]
still no workaround for this problem? I try to clone `VRMHumanoid` but doesn't work
